### DISCUSSION TITLE
How to debug and hotswap code change when developing?

### DISCUSSION THREAD

**Discussion by Amejonah1200** [zed-industries/zed#18205] [Feature Requests]
Hello o7
I wanted to ask about the topic of hot swap, notably while developing zed itself. I wanted to test things out in zed's code but then noticed that... to test those changes, I need to recompile like 20 dependent modules each time and I can't just `ctrl+s` and see my changes reflected...

How do you do that, zed team and others? How do you develop zed while keeping a fast feedback loop? Or are you just coping with this restart-recompile situation?

**Comment by SomeoneToIgnore** [comment]
More correct question to ask would be "Does Rust have a working hot swap functionality?"
I am only aware of https://github.com/draivin/rust-hotswap which seems abandoned.

**Reply by Amejonah1200** [reply]
I mean, what is your own tool set, as a maintainer of zed, for working on it? I can't fathom, coming from Kotlin and JVM, on how annoying it must be. Unless you just don't need to do it, I suppose?

**Reply by SomeoneToIgnore** [reply]
I have no specific tools, using Zed and rust-analyzer, cargo, debug prints (+occasional, real debugging for complex cases) and compilation.
No hot swap or anything like that, as you should have guessed by now.
There are tasks in Zed, so rerunning my code is a matter of pressing a single keystroke after saving a file which is very fast to [re]trigger when needed.
IIRC, there's no major differences between workflows of other maintainers and mine.

I do not find this all annoying, as mostly doing `cargo check` without real compilation or rerunning a specific test in a loop: this all is pretty standard in Rust development and I guess I'm used to that.
I have relatively beefy CPUs on my machines, so Zed debug rebuilds are tolerable (albeit could be better).
Also, "like 20 dependent modules" is not always the case unless you edit something relatively deep in the crate graph.
The only notable exception I seem to have compared to others', is a [fully disabled](https://zed.dev/docs/languages/rust#large-projects-and-performance) "cargo check on save" feature in rust-analyzer which I consider quite harmful for large projects.

From my previous JVM-based experience, I remember frustrating compile times from Scala (worse that Rust ones in my perception) and painfully slow startups of Java applications "thanks" to either bloated Spring contexts or horrendous application servers like IBM WebSphere.
Back in my JVM days, _free_ hot swap tools were very limited and were mostly used to change a specific method's body which is not that useful: any Spring config change or new method(s) added forced me to wait for the whole compile & startup pipeline.
I guess the situation is better now there, but still it does not seem to matter for me much.

**Reply by Amejonah1200** [reply]
Thank you for the fast answer by the way, I truly appreciate it!

> were mostly used to change a specific method's body which is not that useful

They can be incredibly useful if dealing with UI in particular. That is the part I would want to more "hacking" the editor.

By potentially trying out stuff to make the UX more like IntelliJ I'm used to.

While I do agree that cargo check or clippy is helpful and mostly will suffice (for backend development, or any non UI-dev for that matter). UI dev it is kinda not really there as I cannot guess the outcome of my change without code running. (for example changing how a text color would be displayed).

> There are tasks in Zed, so rerunning my code is a matter of pressing a single keystroke after saving a file which is very fast to [re]trigger when needed.

Can you please elaborate on that? What are "tasks" in zed's context (I'm new to the codebase)?

How is the workflow within zed?

**Reply by SomeoneToIgnore** [reply]
Have you tried searching what Zed tasks are first?

https://zed.dev/docs/tasks

**Reply by Amejonah1200** [reply]
*A 6-hour sleep and a workday later*

> Have you tried searching what Zed tasks are first?

Not really, but as I have imagined, it is more of a backend, non-UI development. Which of course doesn't need hotswap that much. But UI dev (if I wanted to add some form of pop-up or highlighting fix), it would be a pain to work with, how does the UI "team" handle that?

**Reply by SomeoneToIgnore** [reply]
I guess they tolerate, same as I do when touching the UI.

There's been a prototype of a "file-based" UI for debug builds of Zed, where a person could define some sort of a template and alter it on the fly — yet, after that we've rewritten gpui and the prototype is not that demanded, apparently, so not ported.

**Reply by Amejonah1200** [reply]
Hmm, that ""file-based" UI" sounds as an interesting idea...
Instead of using rust to create the UI, a settings file or alike could be made, where values are tweaked using a JSON file to adjust and test on the fly (using a watcher). Sounds like a compromise in that situation...

Anyway, I think with the disabling of full-project cargo check and some other tweaks, it might become bearable to quickly check out on how some changes might be introduced. Thank you, very much, for your answers!
